### ISSUE TITLE
hard negative mining

### ISSUE THREAD

**Issue by kmsravindra** [AlexeyAB/darknet#1331]
@AlexeyAB , Wanted to know if hard negative mining can be implemented in yolov3 loss functions? [Joseph Redmon's YOLOv3 paper](https://pjreddie.com/media/files/papers/YOLOv3.pdf) simply qoutes "We still train on full images with no hard negative mining
or any of that stuff." without much reasoning as to why that was omitted. 

However, considering YOLO type of architectures where the predictions out number the actual objects in any given frame, we are training lot of background (negative regions) than positive objects creating a class imbalance between background and objects? So isn't hard negative mining important for yolo?

It would be great to know your view if it would be effective if it is added as a post processing step after YOLO detections? OR any other alternatives exist as a workaround to tackle this?

**Comment by AlexeyAB** [comment]
@kmsravindra 

May be there is no high imbalance, due to 
* using separate `objectness` (T0) parameter
* using cross-entropy loss for `objectness` (T0) without logistic gradient (that was used in Yolo v2)

----

Do you mean that Yolo v3 calculates much more `delta`s for background than for objects (is trained more to avoid false-positive than to predict true-positive)?

I think the point is in that Yolo v3 uses `objectness`, so:

* for **backgrounds** it calculates `delta`s only for `objectness` https://github.com/AlexeyAB/darknet/blob/b8e6e80c6d411d05a9e09f1e3676eb9a7f3ea0e8/src/yolo_layer.c#L218

* for **objects** it calculates `delta`s for 
    * `objectness` https://github.com/AlexeyAB/darknet/blob/b8e6e80c6d411d05a9e09f1e3676eb9a7f3ea0e8/src/yolo_layer.c#L265
    * all-classes https://github.com/AlexeyAB/darknet/blob/b8e6e80c6d411d05a9e09f1e3676eb9a7f3ea0e8/src/yolo_layer.c#L270
    * box-coordinates (x,y,w,h) https://github.com/AlexeyAB/darknet/blob/b8e6e80c6d411d05a9e09f1e3676eb9a7f3ea0e8/src/yolo_layer.c#L261

For example, the 1st [yolo]-layer 13x13 x 3(anchors) for training for 80 objects:
* for background-only image it calculates `13*13*3 = 507` non-zero deltas
* for 3 objects it calculates `3 * (4+1+80) = 255` non-zero deltas

So 507/255 = ~2 is not high imbalance.

----

In the Yolo v1 and v2 there were scales to avoid such imbalancing: 
* v1: https://github.com/AlexeyAB/darknet/blob/1b85111217e90bc20907070283d5de9d9d750057/cfg/yolov1/yolo.cfg#L253-L256
* v2 https://github.com/AlexeyAB/darknet/blob/1b85111217e90bc20907070283d5de9d9d750057/cfg/yolov2.cfg#L251-L254

But in the Yolo v3 these scales was removed, my be because it isn't required due to different way - how to calculate `delta` for `objectness` in v2 and v3: https://github.com/AlexeyAB/darknet/issues/559#issuecomment-378086536

**Comment by kmsravindra** [comment]
Thanks @AlexeyAB , while this sounds reasonable there are few things I am not clear upon - 

1. May be for 80 classes, there isn't much imbalance, but in my case I have only 1 class. So maybe setting focal_loss =1 should help? Will give it a try.

2. Apart from background only images, for the images that have an object annotated, isn't the complement of the bounding box area also adding up to background training and thereby increasing the number of deltas for backgrounds?

The reason is that I am getting some good number of false positives on look alike objects (despite training on huge dataset) and I am looking for ways to reduce such false positives...Thought that some hard negative training is the next logical step. It will be helpful if you have any ideas to reduce false positives in my case.

**Comment by AlexeyAB** [comment]
1. Yes, you can try `focal_loss =1`

2. Yes, I mean that on a single image with 3 objects imbalance will be about 1/2

----

> The reason is that I am getting some good number of false positives 

Solving this imbalance problem can decrease false-negatives, but can't decrease false-positives. Quite the contrary,  false-positives will be increased.


To decrease false-positives - you should add negative-samples: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> desirable that your training dataset include images with non-labeled objects that you do not want to detect - negative samples without bounded box (empty .txt files) - use as many images of negative samples as there are images with objects

**Comment by kmsravindra** [comment]
Thanks @AlexeyAB . Though I have added negative samples, I guess I need to add more variety of negative samples. Thanks for your help.

**Comment by kmsravindra** [comment]
@AlexeyAB,

I just finished training yolov3 with focal loss, and found that this focal loss yolov3 improved the mAP by almost 3.5%. compared to my earlier one w/o focal loss!!
Earlier the mAP was around 56% and now it is 59.4% for the same 3000 iterations training and inferred on the same validation set.
Compared to earlier run on validation :- True positives increased significantly, false negatives reduced as well but false positives almost remained same - just as you mentioned...

I am trying to comprehend why did false negatives come down instead of false positives with focal loss? My understanding of focal loss is that that it would push the loss to zero for well classified examples be it background or objectness...so indirectly the network should get trained upon hard data and thereby reduce both false positives and false negatives?
